{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final Project of Introduction to Bioinformatics\n",
    "\n",
    "## Comparative Mitochondrial Genomics and Phylogenetic Analysis\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This task focuses on the comparative analysis of mitochondrial genomes from different species, primarily birds, mammals, and insects. The aim is to understand the evolutionary relationships between these species by analyzing and comparing their mitochondrial DNA, which is about 16,000 base pairs in length. You will use advanced computational methods to construct phylogenetic trees and delve into the ecological and anthropological insights that can be gleaned from this data. This project is designed to provide a comprehensive understanding of mitochondrial genomics, its importance in evolutionary biology, and its applications in broader scientific contexts.\n",
    "\n",
    "You will learn:\n",
    "\n",
    "- Techniques for aligning and comparing mitochondrial DNA sequences.\n",
    "- How to construct and interpret phylogenetic trees using advanced computational methods.\n",
    "- The application of mitochondrial genomics in understanding ecological interactions and human evolutionary history.\n",
    "\n",
    "#### Task Roadmap\n",
    "\n",
    "1. **Mitochondrial Genome Comparison**:\n",
    "   - Align mitochondrial DNA sequences from the provided dataset.\n",
    "   - Analyze these sequences to identify similarities and differences across species.\n",
    "\n",
    "2. **Phylogenetic Analysis Using Advanced Methods**:\n",
    "   - Apply Maximum Likelihood (ML) and Bayesian Inference methods, utilizing tools like `ETE Toolkit`, `DendroPy`, `BEAST`, or `PyRate`.\n",
    "   - Compare the trees generated by these methods to understand how different approaches can lead to different interpretations of the data.\n",
    "\n",
    "3. **Cross-Disciplinary Applications (Bonus)**:\n",
    "   - **Ecology**: Examine how mitochondrial DNA analysis can reveal information about species adaptation, migration, and conservation. This involves understanding how genetic variation within and between species can inform ecological strategies and conservation efforts.\n",
    "   - **Anthropology**: Investigate the use of mitochondrial DNA in tracing human evolution and migration patterns. This includes studying the mitochondrial DNA of mammals in your dataset to draw parallels with human evolutionary studies.\n",
    "\n",
    "### Data Sources\n",
    "\n",
    "The mitochondrial DNA data for birds, mammals, and insects will be provided to you. This dataset has been curated to facilitate a comprehensive comparative analysis and is essential for the completion of your phylogenetic studies.\n",
    "\n",
    "### Useful Resources and Material\n",
    "\n",
    "- [Mitochondrial DNA - Wikipedia](https://en.wikipedia.org/wiki/Mitochondrial_DNA): A general introduction to the structure, function, origin, and diversity of mitochondrial DNA, as well as its applications in various fields such as medicine, forensics, and anthropology.\n",
    "- [Mitochondrial DNA Analysis: Introduction, Methods, and Applications](https://bioinfo.cd-genomics.com/mitochondrial-dna-analysis-introduction-methods-and-applications.html): An explanation of the basics of mitochondrial DNA sequencing, bioinformatics analysis, heteroplasmy, and advantages of mitochondrial DNA analysis over nuclear DNA analysis.\n",
    "- [Phylogenetic Tree- Definition, Types, Steps, Methods, Uses - Microbe Notes](https://microbenotes.com/phylogenetic-tree/): A coverage of the concepts and methods of phylogenetic tree construction, including the types of phylogenetic trees, the steps involved in phylogenetic analysis, the main methods of phylogenetic inference, and the applications of phylogenetic trees in various disciplines.\n",
    "- [Phylogenetics - Wikipedia](https://en.wikipedia.org/wiki/Phylogenetics): An overview of the field of phylogenetics, which is the study of the evolutionary history and relationships among or within groups of organisms. It also discusses the data sources, models, algorithms, software, and challenges of phylogenetic analysis.\n",
    "- [ETE Toolkit](http://etetoolkit.org/): A Python library for manipulating, analyzing, and visualizing phylogenetic trees. It supports various formats, methods, and tools for phylogenetic analysis, such as alignment, tree inference, tree comparison, tree annotation, and tree visualization.\n",
    "- [DendroPy](https://dendropy.org/): Another Python library for phylogenetic computing. It provides a comprehensive API for working with phylogenetic data structures, such as trees, characters, and networks. It also offers a rich set of functions for simulation, manipulation, analysis, and annotation of phylogenetic data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploration and Reflection\n",
    "\n",
    "As we proceed with our analysis of mitochondrial DNA for phylogenetic tree construction, it is valuable to contemplate a few questions. These inquiries aim to facilitate a more thorough understanding of the roles and characteristics of mitochondrial DNA in the context of evolutionary biology:\n",
    "\n",
    "1. **Maternal Inheritance and Its Implications**: How does the maternal inheritance of mitochondrial DNA simplify our understanding of evolutionary lineage compared to nuclear DNA, which undergoes recombination? What unique insights can this aspect provide in tracing the evolutionary history of species?\n",
    "\n",
    "2. **Mutation Rate and Evolutionary Insights**: Mitochondrial DNA mutates at a faster rate than nuclear DNA. How does this characteristic make mtDNA a more sensitive tool for detecting recent evolutionary events and relationships among closely related species? Can you think of any specific scenarios or studies where this property of mtDNA has been particularly instrumental?\n",
    "\n",
    "Reflect on these questions as you work through the project, and consider how the properties of mitochondrial DNA enhance its value and applicability in evolutionary biology and beyond. Provide your answer either in this notebook, or in your report (if you had one).\n",
    "\n",
    "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
    "    <strong>Don't forget to answer these questions!</strong>\n",
    "</blockquote>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 0: Installing Necessary Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biopython is NOT installed\n",
      "biopython was successfully installed.\n",
      "pandas (2.1.3) is installed\n",
      "numpy (1.26.2) is installed\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import subprocess\n",
    "import pkg_resources\n",
    "\n",
    "def install(package):\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
    "\n",
    "REQUIRED_PACKAGES = [\n",
    "    'biopython',\n",
    "    'pandas',\n",
    "    'numpy'\n",
    "]\n",
    "\n",
    "for package in REQUIRED_PACKAGES:\n",
    "    try:\n",
    "        dist = pkg_resources.get_distribution(package)\n",
    "        print('{} ({}) is installed'.format(dist.key, dist.version))\n",
    "    except pkg_resources.DistributionNotFound:\n",
    "        print('{} is NOT installed'.format(package))\n",
    "        install(package)\n",
    "        print('{} was successfully installed.'.format(package))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary libraries. \n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Dataset Expansion\n",
    "\n",
    "Our first task is to augment our dataset with additional species. This involves engaging with the NCBI database to retrieve mitochondrial DNA sequences.\n",
    "\n",
    "#### Instructions:\n",
    "\n",
    "- **Species Selection**: Identify and choose 10 additional species to include in your dataset. Aim for a diverse selection to enrich your phylogenetic analysis.\n",
    "\n",
    "- **Querying NCBI Database**: Use the NCBI database to locate mitochondrial DNA sequences for your chosen species. While you can manually search on the [NCBI website](https://www.ncbi.nlm.nih.gov/), consider automating this process through their API for a more efficient approach.\n",
    "    - **Example Query**: As a starting point, you might use a query like `\"mitochondrion[Filter] AND (your_species_name[Organism])` to find specific mtDNA sequences. Adjust the query parameters according to your species selection.\n",
    "    - **Documentation**: Familiarize yourself with the [NCBI API documentation](https://www.ncbi.nlm.nih.gov/books/NBK25497/) for detailed guidance on constructing queries.\n",
    "\n",
    "- **Using NCBI Website**: You are welcome to use the NCBI website for this task. If you do so, document each step of your process clearly in your task report. This should include the species names, search terms used, and how you determined the relevant sequences to include.\n",
    "\n",
    "- **Bonus Opportunity**: Implementing an automated, methodological approach using the NCBI API and relevant Python packages to add all 10 records in your dataset will earn you a 50% bonus for this section. Your method should be structured and replicable, demonstrating a systematic approach to data collection.\n",
    "\n",
    "Remember, the goal is to methodically expand your dataset with relevant mtDNA sequences, paving the way for insightful phylogenetic analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NC_000012.12\n",
      "Name: NC_000012\n",
      "Description: Homo sapiens chromosome 12, GRCh38.p14 Primary Assembly\n",
      "Database cross-references: BioProject:PRJNA168, Assembly:GCF_000001405.40\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=07-OCT-2023\n",
      "/accessions=['NC_000012']\n",
      "/sequence_version=12\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='The finished DNA sequence of human chromosome 12', ...), Reference(title='Finishing the euchromatic sequence of the human genome', ...), Reference(title='Initial sequencing and analysis of the human genome', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CM000674.2.\n",
      "On Feb 3, 2014 this sequence version replaced NC_000012.11.\n",
      "Assembly Name: GRCh38.p14 Primary Assembly\n",
      "The DNA sequence is composed of genomic sequence, primarily\n",
      "finished clones that were sequenced as part of the Human Genome\n",
      "Project. PCR products and WGS shotgun sequence have been added\n",
      "where necessary to fill gaps or correct errors. All such additions\n",
      "are manually curated by GRC staff. For more information see:\n",
      "https://genomereference.org.\n",
      "/structured_comment=defaultdict(<class 'dict'>, {'Genome-Annotation-Data': {'Annotation Provider': 'NCBI RefSeq', 'Annotation Status': 'Updated annotation', 'Annotation Name': 'GCF_000001405.40-RS_2023_10', 'Annotation Pipeline': 'NCBI eukaryotic genome annotation pipeline', 'Annotation Software Version': '10.2', 'Annotation Method': 'Best-placed RefSeq; Gnomon; RefSeqFE; cmsearch; tRNAscan-SE', 'Features Annotated': 'Gene; mRNA; CDS; ncRNA', 'Annotation Date': '10/02/2023'}})\n",
      "/contig=join(gap(10000),NT_009759.17:1..7073650,gap(1000),NT_009714.18:1..27634757,gap(50000),NT_187222.1:1..47204,gap(100),NT_187223.1:1..3474,gap(100),NT_187224.1:1..11803,gap(100),NT_187225.1:1..3007,gap(100),NT_187226.1:1..2349957,gap(50000),NT_029419.13:1..94988110,gap(1000),NT_024477.15:1..1040947,gap(10000))\n",
      "Undefined sequence of length 133275309\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def download_sequence(id):\n",
    "    Entrez.email = \"your_email@example.com\"  # Always tell NCBI who you are\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=id, rettype=\"gb\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    return record\n",
    "\n",
    "# Replace 'NC_000012.12' with your accession number\n",
    "record = download_sequence('NC_000012.12')\n",
    "\n",
    "# Replace '9606' with your taxonomy ID\n",
    "# taxonomy = download_sequence('9606')\n",
    "print(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object parse at 0x000001E352002B90>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "record = NCBIXML.parse(taxonomy)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_blast_results(blast_records):\n",
    "    \"\"\"\n",
    "    Parse BLAST results and extract relevant information including taxonomy.\n",
    "    \"\"\"\n",
    "    blast_results = []\n",
    "\n",
    "    for record in blast_records:\n",
    "        for alignment in record.alignments:\n",
    "            accession = alignment.accession\n",
    "            taxonomy, species = fetch_taxonomy_info(accession)\n",
    "            for hsp in alignment.hsps:\n",
    "                # These fields are required in your submission\n",
    "                blast_results.append({\n",
    "                    'query_id': record.query_id,\n",
    "                    'alignment_title': alignment.title,\n",
    "                    'e_value': hsp.expect,\n",
    "                    'identity': hsp.identities,\n",
    "                    'accession': accession,\n",
    "                    'taxonomy': taxonomy,\n",
    "                    'species': species\n",
    "                })\n",
    "    return blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxo_id</th>\n",
       "      <th>specie</th>\n",
       "      <th>blast_name</th>\n",
       "      <th>genbank_common_name</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>mtDNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8945</td>\n",
       "      <td>Eudynamys scolopaceus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Asian koel</td>\n",
       "      <td>NC_060520</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7460</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Bees</td>\n",
       "      <td>honey bee</td>\n",
       "      <td>NC_051932</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36300</td>\n",
       "      <td>Pelecanus crispus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Dalmatian pelican</td>\n",
       "      <td>OR620163</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Rodents</td>\n",
       "      <td>Norway rat</td>\n",
       "      <td>NC_001665</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9031</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>NC_053523</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxo_id                 specie blast_name genbank_common_name  \\\n",
       "0     8945  Eudynamys scolopaceus      birds          Asian koel   \n",
       "1     7460         Apis mellifera       Bees           honey bee   \n",
       "2    36300      Pelecanus crispus      birds   Dalmatian pelican   \n",
       "3    10116      Rattus norvegicus    Rodents          Norway rat   \n",
       "4     9031          Gallus gallus      birds       Gallus gallus   \n",
       "\n",
       "  accession_number                                              mtDNA  \n",
       "0        NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
       "1        NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
       "2         OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
       "3        NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
       "4        NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = pd.read_csv('./dataset/species.csv')\n",
    "\n",
    "dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# TODO: Add 10 more species to your dataset\n",
    "# for species in additional_species:\n",
    "    # Fetch mtDNA and add to the dataset\n",
    "\n",
    "# TODO: Save the expanded dataset\n",
    "# dataset.to_csv('./dataset/expanded_dataset.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### My codes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['DbList'])\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = \"alizaden.1379@gmail.com\" \n",
    "handle = Entrez.einfo()\n",
    "rec = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(rec.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec['DbList']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "## functions to get data from NCBI, given specie name\n",
    "# get taxonomy Id from specie name\n",
    "def get_taxonomy_id(species):\n",
    "    search = Entrez.esearch(term=species, db=\"taxonomy\", retmode=\"xml\")\n",
    "    record = Entrez.read(search)\n",
    "    return record['IdList'][0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get blast name and genbank name from tax id:\n",
    "from requests.adapters import HTTPAdapter\n",
    "from urllib3.util.retry import Retry\n",
    "from io import BytesIO\n",
    "import requests\n",
    "\n",
    "def fetch_entrez_record(db, id, rettype, retmode):\n",
    "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
    "    session = requests.Session()\n",
    "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
    "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
    "\n",
    "    try:\n",
    "        response = session.get(url)\n",
    "        response.raise_for_status()\n",
    "        if db == \"nucleotide\":\n",
    "            return response.text\n",
    "        return response.content\n",
    "    except requests.exceptions.HTTPError as err:\n",
    "        print(f\"HTTP error: {err}\")\n",
    "    except requests.exceptions.ConnectionError as err:\n",
    "        print(f\"Connection error: {err}\")\n",
    "    return None\n",
    "\n",
    "def get_blast_genbank_name_from_taxid(taxonomy_id):\n",
    "    xml_data = fetch_entrez_record(\"taxonomy\", taxonomy_id, \"xml\", \"xml\")\n",
    "    if xml_data:\n",
    "        # Convert bytes data to a binary file-like object\n",
    "        xml_data_io = BytesIO(xml_data)\n",
    "        records = Entrez.read(xml_data_io)\n",
    "        if len(records[0]['OtherNames']['CommonName'])>0:\n",
    "            blast_name=records[0]['OtherNames']['CommonName'][0]\n",
    "        else:\n",
    "            blast_name=records[0]['OtherNames']['GenbankCommonName']\n",
    "        genbank_name=records[0]['OtherNames']['GenbankCommonName']\n",
    "        return blast_name,genbank_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get accession number from specie name\n",
    "def get_accession_number(species):\n",
    "    search = Entrez.esearch(term=f\"{species}[Orgn] AND mitochondrion[Filter]\", db=\"nucleotide\", retmode=\"xml\")\n",
    "    record = Entrez.read(search)\n",
    "    id_list = record['IdList']\n",
    "    \n",
    "    if id_list:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=id_list[0], rettype=\"gb\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        return record.id\n",
    "    else:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "## final function to create a row for dataset given specie name\n",
    "def get_info(specie):\n",
    "    tax_id=get_taxonomy_id(specie)\n",
    "    blast_name,genbank_name=get_blast_genbank_name_from_taxid(tax_id)\n",
    "    accession_number=get_accession_number(specie)\n",
    "    link=\"https://www.ncbi.nlm.nih.gov/nuccore/\"+accession_number\n",
    "    return {'taxo_id': tax_id,\n",
    "            'specie': specie,\n",
    "            'blast_name':blast_name,\n",
    "            'genbank_common_name':genbank_name,\n",
    "            'accession_number':accession_number.split('.')[0],\n",
    "            'mtDNA':link}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# species are cat, dog, fox, butterfly, mouse, ladybug, pigeon, chicken, asian elephant, tiger\n",
    "species=['Felis catus','Canis lupus','Vulpes vulpes','Lepidoptera','Mus musculus','Coccinellidae','Columba livia','Gallus gallus domesticus','Elephas maximus','Panthera tigris']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_dataset=dataset.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxo_id</th>\n",
       "      <th>specie</th>\n",
       "      <th>blast_name</th>\n",
       "      <th>genbank_common_name</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>mtDNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8945</td>\n",
       "      <td>Eudynamys scolopaceus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Asian koel</td>\n",
       "      <td>NC_060520</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7460</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Bees</td>\n",
       "      <td>honey bee</td>\n",
       "      <td>NC_051932</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36300</td>\n",
       "      <td>Pelecanus crispus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Dalmatian pelican</td>\n",
       "      <td>OR620163</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Rodents</td>\n",
       "      <td>Norway rat</td>\n",
       "      <td>NC_001665</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9031</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>NC_053523</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>59463</td>\n",
       "      <td>Myotis lucifugus</td>\n",
       "      <td>bats</td>\n",
       "      <td>little brown bat</td>\n",
       "      <td>NC_029849</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7070</td>\n",
       "      <td>Tribolium castaneum</td>\n",
       "      <td>Insects</td>\n",
       "      <td>Tribolium castaneum</td>\n",
       "      <td>NC_003081</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_003081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8845</td>\n",
       "      <td>Anser cygnoides</td>\n",
       "      <td>birds</td>\n",
       "      <td>Swan goose</td>\n",
       "      <td>NC_023832</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9940</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>sheep</td>\n",
       "      <td>even-toed ungulates</td>\n",
       "      <td>NC_001941</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>56313</td>\n",
       "      <td>Tyto alba</td>\n",
       "      <td>birds</td>\n",
       "      <td>common barn owl</td>\n",
       "      <td>MZ318036</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>flies</td>\n",
       "      <td>fruit fly</td>\n",
       "      <td>NC_024511</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>8822</td>\n",
       "      <td>Apteryx australis</td>\n",
       "      <td>birds</td>\n",
       "      <td>Southern brown kiwi</td>\n",
       "      <td>MN356385</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>92683</td>\n",
       "      <td>Spheniscus demersus</td>\n",
       "      <td>birds</td>\n",
       "      <td>jackass penguin</td>\n",
       "      <td>NC_022817</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>85066</td>\n",
       "      <td>Corvus brachyrhynchos</td>\n",
       "      <td>birds</td>\n",
       "      <td>American crow</td>\n",
       "      <td>NC_026461</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>rodents</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>NC_005089</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_005089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8790</td>\n",
       "      <td>Dromaius novaehollandiae</td>\n",
       "      <td>birds</td>\n",
       "      <td>emu</td>\n",
       "      <td>NC_002784</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>8839</td>\n",
       "      <td>Anas platyrhynchos</td>\n",
       "      <td>birds</td>\n",
       "      <td>mallard</td>\n",
       "      <td>NC_009684</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_009684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>132908</td>\n",
       "      <td>Pteropus vampyrus</td>\n",
       "      <td>bats</td>\n",
       "      <td>large flying fox</td>\n",
       "      <td>NC_026542</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>8801</td>\n",
       "      <td>Struthio camelus</td>\n",
       "      <td>birds</td>\n",
       "      <td>African ostrich</td>\n",
       "      <td>NC_002785</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>193005</td>\n",
       "      <td>Aquila nipalensis</td>\n",
       "      <td>hawks &amp; eagles</td>\n",
       "      <td>steppe eagle</td>\n",
       "      <td>NC_045042</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>219595</td>\n",
       "      <td>Cygnus cygnus</td>\n",
       "      <td>birds</td>\n",
       "      <td>common whooper</td>\n",
       "      <td>NC_027095</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>59729</td>\n",
       "      <td>Taeniopygia guttata</td>\n",
       "      <td>birds</td>\n",
       "      <td>zebra finch</td>\n",
       "      <td>NC_007897</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>9233</td>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>birds</td>\n",
       "      <td>emperor penguin</td>\n",
       "      <td>NC_027938</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>9685</td>\n",
       "      <td>Felis catus</td>\n",
       "      <td>cat</td>\n",
       "      <td>domestic cat</td>\n",
       "      <td>NC_001700</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001700.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>9612</td>\n",
       "      <td>Canis lupus</td>\n",
       "      <td>grey wolf</td>\n",
       "      <td>gray wolf</td>\n",
       "      <td>NC_010340</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_010340.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>9627</td>\n",
       "      <td>Vulpes vulpes</td>\n",
       "      <td>silver fox</td>\n",
       "      <td>red fox</td>\n",
       "      <td>MT941681</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MT941681.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>7088</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>moths</td>\n",
       "      <td>butterflies and moths</td>\n",
       "      <td>KU965151</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/KU965151.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>mouse</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>KY986857</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/KY986857.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>7080</td>\n",
       "      <td>Coccinellidae</td>\n",
       "      <td>lady beetles</td>\n",
       "      <td>ladybird beetles</td>\n",
       "      <td>MW392901</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MW392901.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>8932</td>\n",
       "      <td>Columba livia</td>\n",
       "      <td>carrier pigeon</td>\n",
       "      <td>rock pigeon</td>\n",
       "      <td>NC_013978</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_013978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>9031</td>\n",
       "      <td>Gallus gallus domesticus</td>\n",
       "      <td>bantam</td>\n",
       "      <td>chicken</td>\n",
       "      <td>NC_053523</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>9783</td>\n",
       "      <td>Elephas maximus</td>\n",
       "      <td>Asian elephant</td>\n",
       "      <td>Asiatic elephant</td>\n",
       "      <td>MW703462</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MW703462.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>9694</td>\n",
       "      <td>Panthera tigris</td>\n",
       "      <td>tiger</td>\n",
       "      <td>tiger</td>\n",
       "      <td>NC_014770</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_014770.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxo_id                    specie      blast_name    genbank_common_name  \\\n",
       "0     8945     Eudynamys scolopaceus           birds             Asian koel   \n",
       "1     7460            Apis mellifera            Bees              honey bee   \n",
       "2    36300         Pelecanus crispus           birds      Dalmatian pelican   \n",
       "3    10116         Rattus norvegicus         Rodents             Norway rat   \n",
       "4     9031             Gallus gallus           birds          Gallus gallus   \n",
       "5    59463          Myotis lucifugus            bats       little brown bat   \n",
       "6     7070       Tribolium castaneum         Insects    Tribolium castaneum   \n",
       "7     8845           Anser cygnoides           birds             Swan goose   \n",
       "8     9940                Ovis aries           sheep    even-toed ungulates   \n",
       "9    56313                 Tyto alba           birds        common barn owl   \n",
       "10    7227   Drosophila melanogaster           flies              fruit fly   \n",
       "11    8822         Apteryx australis           birds    Southern brown kiwi   \n",
       "12   92683       Spheniscus demersus           birds        jackass penguin   \n",
       "13   85066     Corvus brachyrhynchos           birds          American crow   \n",
       "14   10090              Mus musculus         rodents            house mouse   \n",
       "15    8790  Dromaius novaehollandiae           birds                    emu   \n",
       "16    8839        Anas platyrhynchos           birds                mallard   \n",
       "17  132908         Pteropus vampyrus            bats       large flying fox   \n",
       "18    8801          Struthio camelus           birds        African ostrich   \n",
       "19  193005         Aquila nipalensis  hawks & eagles           steppe eagle   \n",
       "20  219595             Cygnus cygnus           birds         common whooper   \n",
       "21   59729       Taeniopygia guttata           birds            zebra finch   \n",
       "22    9233      Aptenodytes forsteri           birds        emperor penguin   \n",
       "23    9685               Felis catus             cat           domestic cat   \n",
       "24    9612               Canis lupus       grey wolf              gray wolf   \n",
       "25    9627             Vulpes vulpes      silver fox                red fox   \n",
       "26    7088               Lepidoptera           moths  butterflies and moths   \n",
       "27   10090              Mus musculus           mouse            house mouse   \n",
       "28    7080             Coccinellidae    lady beetles       ladybird beetles   \n",
       "29    8932             Columba livia  carrier pigeon            rock pigeon   \n",
       "30    9031  Gallus gallus domesticus          bantam                chicken   \n",
       "31    9783           Elephas maximus  Asian elephant       Asiatic elephant   \n",
       "32    9694           Panthera tigris           tiger                  tiger   \n",
       "\n",
       "   accession_number                                              mtDNA  \n",
       "0         NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
       "1         NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
       "2          OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
       "3         NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
       "4         NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  \n",
       "5         NC_029849   https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1  \n",
       "6         NC_003081     https://www.ncbi.nlm.nih.gov/nuccore/NC_003081  \n",
       "7         NC_023832  https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...  \n",
       "8         NC_001941     https://www.ncbi.nlm.nih.gov/nuccore/NC_001941  \n",
       "9          MZ318036    https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1  \n",
       "10        NC_024511  https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...  \n",
       "11         MN356385    https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1  \n",
       "12        NC_022817  https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...  \n",
       "13        NC_026461  https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...  \n",
       "14        NC_005089     https://www.ncbi.nlm.nih.gov/nuccore/NC_005089  \n",
       "15        NC_002784     https://www.ncbi.nlm.nih.gov/nuccore/NC_002784  \n",
       "16        NC_009684     https://www.ncbi.nlm.nih.gov/nuccore/NC_009684  \n",
       "17        NC_026542   https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1  \n",
       "18        NC_002785     https://www.ncbi.nlm.nih.gov/nuccore/NC_002785  \n",
       "19        NC_045042   https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1  \n",
       "20        NC_027095  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  \n",
       "21        NC_007897  https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...  \n",
       "22        NC_027938  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  \n",
       "23        NC_001700   https://www.ncbi.nlm.nih.gov/nuccore/NC_001700.1  \n",
       "24        NC_010340   https://www.ncbi.nlm.nih.gov/nuccore/NC_010340.2  \n",
       "25         MT941681    https://www.ncbi.nlm.nih.gov/nuccore/MT941681.1  \n",
       "26         KU965151    https://www.ncbi.nlm.nih.gov/nuccore/KU965151.1  \n",
       "27         KY986857    https://www.ncbi.nlm.nih.gov/nuccore/KY986857.1  \n",
       "28         MW392901    https://www.ncbi.nlm.nih.gov/nuccore/MW392901.1  \n",
       "29        NC_013978   https://www.ncbi.nlm.nih.gov/nuccore/NC_013978.1  \n",
       "30        NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  \n",
       "31         MW703462    https://www.ncbi.nlm.nih.gov/nuccore/MW703462.1  \n",
       "32        NC_014770   https://www.ncbi.nlm.nih.gov/nuccore/NC_014770.1  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for specie in species:\n",
    "    info=get_info(specie)\n",
    "    expanded_dataset.loc[len(expanded_dataset)]=info\n",
    "expanded_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_dataset.to_csv('./dataset/expanded_dataset.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset=expanded_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Checking Data Consistency\n",
    "\n",
    "The code block is designed to check the accuracy of a biological dataset. It examines each entry in a CSV file, focusing on taxonomy IDs, species names, GenBank accession numbers, and mitochondrial DNA links. It uses the NCBI's Entrez system to ensure taxonomy IDs correspond to the correct species and confirms the mitochondrial DNA links are accurate. The script also checks GenBank accession numbers against the provided links. This method is useful for maintaining the accuracy of current data and **might help in adding new entries to the database.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from Bio import Entrez\n",
    "import os\n",
    "from io import BytesIO\n",
    "from requests.adapters import HTTPAdapter\n",
    "from urllib3.util.retry import Retry\n",
    "\n",
    "# Set your email and API key for NCBI\n",
    "entrez_email = os.getenv('ENTREZ_EMAIL')  \n",
    "entrez_key = os.getenv('ENTREZ_API_KEY')\n",
    "Entrez.email = entrez_email \n",
    "\n",
    "def fetch_entrez_record(db, id, rettype, retmode):\n",
    "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
    "    session = requests.Session()\n",
    "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
    "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
    "\n",
    "    try:\n",
    "        response = session.get(url)\n",
    "        response.raise_for_status()\n",
    "        if db == \"nucleotide\":\n",
    "            return response.text\n",
    "        return response.content\n",
    "    except requests.exceptions.HTTPError as err:\n",
    "        print(f\"HTTP error: {err}\")\n",
    "    except requests.exceptions.ConnectionError as err:\n",
    "        print(f\"Connection error: {err}\")\n",
    "    return None\n",
    "\n",
    "def verify_taxonomy_id(taxo_id, species_name):\n",
    "    \"\"\"Verify taxonomy ID against species name.\"\"\"\n",
    "    xml_data = fetch_entrez_record(\"taxonomy\", taxo_id, \"xml\", \"xml\")\n",
    "    if xml_data:\n",
    "        # Convert bytes data to a binary file-like object\n",
    "        xml_data_io = BytesIO(xml_data)\n",
    "        records = Entrez.read(xml_data_io)\n",
    "        return records[0]['ScientificName'].lower() == species_name.lower()\n",
    "    return False\n",
    "\n",
    "def verify_mitochondrial_dna(accession_number):\n",
    "    gb_data = fetch_entrez_record(\"nucleotide\", accession_number, \"gb\", \"text\")\n",
    "    return \"mitochondrion\" in gb_data.lower() if gb_data else False\n",
    "\n",
    "def extract_accession_from_link(link):\n",
    "    return link.split('/')[-1].split('.')[0]\n",
    "\n",
    "def check_dataset_consistency(file_path):\n",
    "    species_df = pd.read_csv(file_path)\n",
    "\n",
    "    for index, row in species_df.iterrows():\n",
    "        taxonomy_id = str(row['taxo_id'])\n",
    "        species_name = row['specie']\n",
    "        accession_number = row['accession_number']\n",
    "        mtDNA_link = row['mtDNA']\n",
    "        extracted_accession = extract_accession_from_link(mtDNA_link)\n",
    "\n",
    "        taxonomy_check = verify_taxonomy_id(taxonomy_id, species_name)\n",
    "        accession_match = (accession_number == extracted_accession)\n",
    "        mitochondrial_check = verify_mitochondrial_dna(accession_number)\n",
    "\n",
    "        print(f\"Row {index}: Taxonomy Check: {taxonomy_check}, Accession Match: {accession_match}, Mitochondrial Check: {mitochondrial_check}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row 0: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 1: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 2: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 3: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 4: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "HTTP error: 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id=59463&rettype=xml&retmode=xml\n",
      "Row 5: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
      "Row 6: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 7: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 8: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 9: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 10: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 11: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 12: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 13: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 14: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 15: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 16: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 17: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 18: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 19: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 20: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 21: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 22: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 23: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 24: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 25: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 26: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 27: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 28: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 29: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "HTTP error: 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id=9031&rettype=xml&retmode=xml\n",
      "Row 30: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
      "Row 31: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 32: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n"
     ]
    }
   ],
   "source": [
    "check_dataset_consistency('./dataset/expanded_dataset.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3: Sequence Download and Preparation\n",
    "\n",
    "The next step in our project involves downloading the mitochondrial DNA sequences for each species and preparing them for analysis.\n",
    "\n",
    "#### Instructions:\n",
    "\n",
    "- **Download mtDNA Sequences**: Write a script to download the mtDNA sequences from the links provided in your dataset. The sequences should be in FASTA format, which is the standard for nucleotide sequences.\n",
    "\n",
    "- **Sequence Labeling**: Properly label each sequence within the FASTA file. This header, starting with '>', should include the species name and any other relevant information (e.g., `>Eudynamys_scolopaceus_NC_060520`). This is crucial for identifying the sequences in subsequent analysis.\n",
    "\n",
    "- **Concatenate Sequences**:\n",
    "    - Create a script to concatenate all downloaded sequences into a single `.fasta` or `.fna` file. \n",
    "    - Ensure each sequence in the file is clearly separated by its header line, which is important for differentiating the sequences of various species.\n",
    "\n",
    "#### Tips for Writing the Download and Concatenation Script:\n",
    "- Use Python libraries such as `httpx` or `requests`, or any other tool you prefer for downloading sequences. For processing FASTA files you can use a wide range of tools. One recommended option is `Biopython` library.\n",
    "- Use a loop to go through each link in the dataset, download the sequence, and append it to your concatenated file.\n",
    "- Maintain the format integrity of the FASTA file, ensuring each sequence is correctly associated with its header.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Write a function to download mtDNA sequences\n",
    "# def download_mtDNA(url, label):\n",
    "    # Code to download and label the sequence\n",
    "from Bio import Entrez\n",
    "\n",
    "def download_sequence(accession_number):\n",
    "    Entrez.email = \"your_email@example.com\"\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"fasta\", retmode=\"text\")\n",
    "    seq = handle.read()\n",
    "    return seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# TODO: Loop through the dataset and download each mtDNA sequence\n",
    "file=open('Sequences.fasta','w')\n",
    "for index, row in dataset.iterrows():\n",
    "    accession_number=row['accession_number']\n",
    "    sequence=download_sequence(accession_number)\n",
    "    file.write(sequence)\n",
    "\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 4: Sequence Alignment\n",
    "\n",
    "After downloading the mitochondrial DNA sequences, the next critical step is their alignment. This process allows us to compare the sequences and discern the evolutionary relationships among the species.\n",
    "\n",
    "#### Instructions:\n",
    "\n",
    "- **Select an Alignment Tool**: Choose one of the following alignment tools based on your project needs. Each tool has its strengths and is widely used in bioinformatics for multiple sequence alignment.\n",
    "\n",
    "1. **MAFFT**:\n",
    "    - **Brief Introduction**: MAFFT (Multiple Alignment using Fast Fourier Transform) is renowned for its speed and efficiency, particularly suitable for large datasets.\n",
    "    - **Resources**:\n",
    "        - [MAFFT Official Documentation](https://mafft.cbrc.jp/alignment/software/)\n",
    "        - [Example Usage on GitHub](https://github.com/MountainMan12/SARS-Cov2-phylo)\n",
    "        - [Relevant Notebook](https://colab.research.google.com/github/pb3lab/ibm3202/blob/master/tutorials/lab03_phylo.ipynb)\n",
    "\n",
    "2. **Clustal Omega**:\n",
    "    - **Brief Introduction**: Clustal Omega offers high-quality alignments and is user-friendly, ideal for those new to sequence alignment.\n",
    "    - **Resources**:\n",
    "        - [A Python wrapper around Clustal Omega](https://github.com/benchling/clustalo-python)\n",
    "        - [Clustal Omega Official Website](http://www.clustal.org/omega/)\n",
    "\n",
    "3. **MUSCLE**:\n",
    "    - **Brief Introduction**: MUSCLE (Multiple Sequence Comparison by Log-Expectation) is known for its balance between speed and accuracy, making it a versatile choice for various datasets.\n",
    "    - **Resources**:\n",
    "        - [MUSCLE Documentation](https://drive5.com/muscle5/manual/)\n",
    "\n",
    "- **Perform Sequence Alignment**: Utilize your chosen tool to align the downloaded mtDNA sequences. This alignment is foundational for the accurate construction of phylogenetic trees.\n",
    "\n",
    "- **Save Aligned Sequences**: After alignment, save the output in an appropriate format for further analysis in the subsequent steps of this project."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MAFFT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Install and import the alignment tool\n",
    "# !pip install mafft\n",
    "# from Bio.Align.Applications import MafftCommandline\n",
    "from Bio.Align.Applications import MafftCommandline,ClustalOmegaCommandline\n",
    "\n",
    "# TODO: Perform the sequence alignment\n",
    "# def perform_alignment(input_file, output_file):\n",
    "    # Code to align sequences using the chosen tool\n",
    "def perform_alignment(input_file, output_file):\n",
    "    mafft_cline = MafftCommandline(input=input_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    # clustalomega_cline = ClustalOmegaCommandline(infile=input_file, outfile=output_file, verbose=True, auto=True)\n",
    "    # stdout, stderr = clustalomega_cline()\n",
    "    \n",
    "    output=open(output_file, \"w\")\n",
    "    output.write(stdout)\n",
    "    output.close()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 1 from 'mafft Sequences.fasta', message \"'mafft' is not recognized as an internal or external command,\"",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[21], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# TODO: Align your downloaded sequences\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;66;03m# perform_alignment('path_to_downloaded_sequences.fasta', 'aligned_sequences.fasta')\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m \u001b[43mperform_alignment\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mSequences.fasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mAligned_sequences.fasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[20], line 11\u001b[0m, in \u001b[0;36mperform_alignment\u001b[1;34m(input_file, output_file)\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mperform_alignment\u001b[39m(input_file, output_file):\n\u001b[0;32m     10\u001b[0m     mafft_cline \u001b[38;5;241m=\u001b[39m MafftCommandline(\u001b[38;5;28minput\u001b[39m\u001b[38;5;241m=\u001b[39minput_file)\n\u001b[1;32m---> 11\u001b[0m     stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mmafft_cline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     12\u001b[0m     \u001b[38;5;66;03m# clustalomega_cline = ClustalOmegaCommandline(infile=input_file, outfile=output_file, verbose=True, auto=True)\u001b[39;00m\n\u001b[0;32m     13\u001b[0m     \u001b[38;5;66;03m# stdout, stderr = clustalomega_cline()\u001b[39;00m\n\u001b[0;32m     15\u001b[0m     output\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mopen\u001b[39m(output_file, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32md:\\applications\\python\\Lib\\site-packages\\Bio\\Application\\__init__.py:592\u001b[0m, in \u001b[0;36mAbstractCommandline.__call__\u001b[1;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[0;32m    589\u001b[0m     stderr_arg\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m    591\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_code:\n\u001b[1;32m--> 592\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ApplicationError(return_code, \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m), stdout_str, stderr_str)\n\u001b[0;32m    593\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m stdout_str, stderr_str\n",
      "\u001b[1;31mApplicationError\u001b[0m: Non-zero return code 1 from 'mafft Sequences.fasta', message \"'mafft' is not recognized as an internal or external command,\""
     ]
    }
   ],
   "source": [
    "# TODO: Align your downloaded sequences\n",
    "# perform_alignment('path_to_downloaded_sequences.fasta', 'aligned_sequences.fasta')\n",
    "perform_alignment(\"Sequences.fasta\", \"Aligned_sequences.fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 5: Phylogenetic Tree Construction\n",
    "\n",
    "The next phase in our project involves constructing phylogenetic trees to visualize and analyze the evolutionary relationships among the species. We will use three distinct methods, each providing unique insights.\n",
    "\n",
    "#### Phylogenetic Tree Construction Methods:\n",
    "\n",
    "1. **Bayesian Inference Trees**:\n",
    "    - **Overview**: This method uses Bayesian statistics to estimate the likelihood of different evolutionary histories. It's particularly useful for its ability to estimate branch lengths and support values.\n",
    "    - **Tools**: MrBayes, BEAST\n",
    "        - MrBayes ([Official Website](https://nbisweden.github.io/MrBayes/manual.html/)) is widely recognized for its robustness in Bayesian inference.\n",
    "        - BEAST2 ([BEAST Software](https://www.beast2.org/)) is another powerful tool, offering advanced features for complex evolutionary models.\n",
    "\n",
    "2. **Maximum Likelihood Trees**:\n",
    "    - **Overview**: Maximum Likelihood methods evaluate tree topologies based on the likelihood of observed data given a tree model. It's known for its statistical rigor and accuracy.\n",
    "    - **Tools**: RAxML, PhyML\n",
    "        - RAxML ([RAxML GitHub](https://github.com/stamatak/standard-RAxML)) is preferred for large datasets due to its efficiency.\n",
    "        - PhyML ([PhyML Documentation](http://www.atgc-montpellier.fr/phyml/)) offers a balance of speed and accuracy, with a user-friendly interface.\n",
    "\n",
    "3. **Neighbor-Joining Trees**:\n",
    "    - **Overview**: The Neighbor-Joining method is a distance-based approach that constructs phylogenetic trees by evaluating the genetic distance between sequences. It is known for its speed and simplicity, making it well-suited for initial exploratory analyses.\n",
    "    - **Tools**: \n",
    "        - MEGA: A versatile tool specifically used here for constructing Neighbor-Joining trees. It's recognized for its ease of use and effectiveness in phylogenetic analysis. [MEGA Software](https://www.megasoftware.net/)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Import necessary libraries for tree construction\n",
    "# from Bio import Phylo\n",
    "# from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "from Bio import Phylo, AlignIO\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "\n",
    "\n",
    "# TODO: Construct a phylogenetic tree using the Neighbor Joining method\n",
    "# def construct_tree_NJ(aligned_sequences):\n",
    "    # Implement the tree construction using Neighbor Joining\n",
    "def construct_tree_NJ(input_file, output_file):\n",
    "    alignment = AlignIO.read(input_file, \"fasta\")\n",
    "    calculator = DistanceCalculator('identity')\n",
    "    dm = calculator.get_distance(alignment)\n",
    "    constructor = DistanceTreeConstructor()\n",
    "    tree = constructor.nj(dm)\n",
    "    Phylo.write(tree, output_file, 'newick')\n",
    "\n",
    "construct_tree_NJ(\"aligned_sequences.fasta\", \"trees/nj-tree.newick\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def construct_tree_BI(gene_trees):\n",
    "    # Implement the construction of a tree using bayes inference\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Repeat the process for Maximum Likelihood and Supertree methods\n",
    "# def construct_tree_ML(aligned_sequences):\n",
    "    # Implement the tree construction using Maximum Likelihood\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Visualize and save the constructed trees\n",
    "# Phylo.draw(tree, do_show=False)\n",
    "# Phylo.write(tree, 'tree_output.xml', 'phyloxml')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 6: In-Depth Phylogenetic Tree Visualization\n",
    "\n",
    "Having constructed phylogenetic trees using different methods, our next task is to visualize these trees effectively. This step is crucial for interpreting the results and communicating our findings.\n",
    "\n",
    "#### Visualization Tools:\n",
    "\n",
    "1. **FigTree**:\n",
    "    - **Overview**: FigTree is designed for the graphical representation of phylogenetic trees. It's excellent for creating publication-ready visualizations.\n",
    "    - **Resource**: [FigTree Tool](http://tree.bio.ed.ac.uk/software/figtree/)\n",
    "    - **Usage**: Use FigTree to add detailed annotations, adjust branch colors, and format tree layouts for clear, interpretable visualizations.\n",
    "\n",
    "2. **iTOL (Interactive Tree Of Life)**:\n",
    "    - **Overview**: iTOL is a web-based tool for the display, annotation, and management of phylogenetic trees, offering extensive customization options.\n",
    "    - **Resource**: [iTOL Website](https://itol.embl.de/)\n",
    "    - **Usage**: Ideal for interactive tree visualizations. It allows users to explore different layers of data through their tree, such as adding charts or color-coding branches.\n",
    "\n",
    "3. **Dendroscope**:\n",
    "    - **Overview**: Dendroscope is a software program for viewing and editing phylogenetic trees, particularly useful for large datasets.\n",
    "    - **Resource**: [Dendroscope Download](https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/dendroscope/)\n",
    "    - **Usage**: Utilize Dendroscope when dealing with large and complex trees or when you need to compare multiple trees side-by-side.\n",
    "\n",
    "#### Task:\n",
    "\n",
    "- **Visualize Each Tree**: Use one or more of the above tools to visualize the phylogenetic trees you constructed using Bayesian inference, maximum likelihood, and neighbor-joining methods.\n",
    "- **Highlight Differences**: Focus on highlighting the differences and similarities between the trees obtained from the different methods. Pay attention to tree topology, branch lengths, and any notable patterns.\n",
    "- **Interpretation and Presentation**: Aim for visualizations that are not only accurate but also interpretable and visually appealing. This will enhance the clarity of your work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TODO ####\n",
    "# You code for visualization of each tree\n",
    "### TODO ###\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cross-Disciplinary Applications (Optional)\n",
    "\n",
    "This is an optional part with bonus, relative to the depth of your analysis. Refer to the first part of this notebook. You have complete freedom to do this part anyway you like, but to gain a portion of the bonus score for this section, a bare minimum effort is required."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusion and Reflective Insights\n",
    "\n",
    "As we conclude our exploration of phylogenetic tree construction and analysis, let's reflect on the insights learned from this task and consider questions that emerge from our findings.\n",
    "\n",
    "#### Interpretation of Results:\n",
    "\n",
    "- Reflect on the phylogenetic trees produced by each method (Bayesian inference, maximum likelihood, and neighbor-joining). Consider how the differences in tree topology might offer varied perspectives on the evolutionary relationships among the species.\n",
    "\n",
    "#### Questions to Ponder:\n",
    "\n",
    "1. **Species Divergence**: Based on the trees, which species appear to have the most ancient divergence? How might this information contribute to our understanding of their evolutionary history?\n",
    "   \n",
    "2. **Common Ancestors**: Are there any unexpected pairings or groupings of species that suggest a closer evolutionary relationship than previously thought? How could this reshape our understanding of these species' evolutionary paths?\n",
    "\n",
    "3. **Methodology Insights**: Considering the discrepancies between the trees generated by different methods, what might this tell us about the limitations and strengths of each phylogenetic analysis method?\n",
    "\n",
    "4. **Conservation Implications**: Considering the evolutionary relationships revealed in your phylogenetic analysis, what insights can be gained for conservation strategies? Specifically, how could understanding the close evolutionary ties between species, which might be facing distinct environmental challenges, guide targeted conservation efforts?\n",
    "\n",
    "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
    "    <strong>Don't forget to answer these questions!</strong>\n",
    "</blockquote>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PyCharm (open3d-testing)",
   "language": "python",
   "name": "pycharm-ffc329a7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
